Updating Signal interface and adding SignalSum to enable algebraic operations on Signals

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

This PR introduces the `SignalSum` object, containing a list of `Signal` instances and representing their sum. The purpose of this is to naturally preserve frequency information compartmentalization - i.e. if two different `Signal` instances are added together, they still remain internally separated in `SignalSum`, so that frequency-dependent operations/transformations (like RWA) can still analyze them separately. An additional `DiscreteSignalSum` object is introduced, which represents a sum of `DiscreteSignal`s that have compatible sample structure; i.e. the same `dt`, `duration` and `start_time`. These both subclass `Signal`, to inherit much if the same functionality.

Using `SignalSum`, multiplication of two `Signal` instances is implemented to return a `SignalSum` that expands out the multiplication into two terms: one whose frequency is the sum of the frequencies of the product terms, and one which is the difference. Multiplication also has some `Signal`-type aware behavior: e.g. if the two signals being multiplied are `DiscreteSignal` with compatible sample structure, it will implement the multiplication via multiplication of the sample arrays. This is in contrast to the general case: two arbitrarily user-defined `Signal`s, whose envelopes must be multiplied as functions.

Additionally, this PR makes modifications to the original `Signal` interface, the biggest of which is moving to vectorized evaluation.

## Details

`Signal` interface
- Main methods:
  - `envelope(t)` - Vectorized evaluation of the complex envelope
  - `complex_value(t)` - Vectorized evaluation of `f(t)exp(i(2 * pi * v * t + phi))`
  - `__call__(t)` - Vectorized evaluation of `Re[f(t)exp(i(2 * pi * v * t + phi))]`
  - `conjugate()` - Return the signal with the complex conjugate of `f(t)` and phase `-phi`
  - `plot`, `plot_envelope`, `plot_complex_value` - plotting functions
- Main attributes:
  - `carrier_freq` - returns carrier freq
  - `phase` - returns phase
- This class also contains dunder methods for algebraic operations

Note: removed `value` in favour of `__call__`, and changed `envelope_value` to `envelope`.

Changed `PiecewiseConstant` to `DiscreteSignal`.

`SignalSum(Signal)` interface
- Class representing a sum of any subclass of `Signal` (aside from `SignalSum` itself).
- Evaluation functions are modified naturally for a "sum":
  - `envelope` returns an array for each component in the sum for every time
  - `complex_value` and `__call__` return the sum of all signals in the sum.
- Additional attributes:
  - `components` - storage of the list signals in the sum.
- Additional methods:
  - `flatten()` - collapse the sum into a single `Signal`. This destroys frequency information so should be used selectively. This is particularly useful initially as none of the rest of the code has been written to support `SignalSum`s.
  - `__getitem__(idx)` - returns the signal at the `idx` location in the sum. Based on input, this has been written to enable numpy-style indexing. E.g. `idx` can be an integer, a slice, a list of integers, or an array of integers.
  - `__len__` - number of terms in the sum
  - `draw(t0, tf, N, function, axis)` - plot the signal. The `function` option specifies which "function" to plot: i.e. the full signal, the complex envelope, or the complex value.

`DiscreteSignalSum(SignalSum, DiscreteSignal)`
- Class representing a sum of `DiscreteSignal` instances, all of which have the same sample structure: same `dt`, same `duration`, and same `start_time`.
- Implements same interface as `SignalSum` with added `DiscreteSignal` attributes.
- Stores envelope samples for all components in a single array, and all envelopes are evaluated simultaneously with a single array look up.

`SignalList`
- This replaces `VectorSignal`, and is meant to store a list of `Signal`s (and any instances of its subclasses). This has more of a limited interface and currently I only expect this to be used internally.
- `complex_value` - evaluate the complex value (this returns a vector for each element in the list)
- `__call__` - evaluate the list.
- `flatten` - flattens the components.
- `drift` - Evaluates only the constant components.
- `__getitem__` indexing - similar to `SignalSum`, is set up to do fancy indexing.
- `__len__` number of components.

Additional note:
- The `to_pwc` method of `Signal` has been removed, and the functionality has been moved into the constructor `DiscreteSignal.from_Signal`. A similar static method works for `SignalSum`s: `DiscreteSignalSum.from_SignalSum`.

### Algebraic operations implementation

Algebraic operations are implemented via dunder methods in the base `Signal` class:
- `__add__`, `__radd__`, `__mul__`, `__rmul__`, `__neg__`, `__sub__`, and `__rsub__`
- Where relevant, the above operations also work with scalars (they get wrapped in `Constant`s)
- Addition is implemented via direct construction of a `SignalSum` object
  - If one (or both) of the signals being added together are a `SignalSum`, they will be appropriately unrolled.
- Multiplication is performed by constructing a `SignalSum` via the relation:
  - `Re[f(t)exp(i(wt + phi))] * Re[g(t)exp(i(vt + psi))] =` 
       `Re[0.5 *f(t)g(t)exp(i( (w+v)t + (phi + psi)))] + Re[0.5 *f(t)conj(g(t))exp(i( (w-v)t + (phi - psi)))]`
  - The above relation works for `SignalSum`s by performing the above operation for all pairs in the sums (distributivity of multiplication over sums).
- _Multiplication of envelopes_: the above formula involves multiplication of envelopes. Where possible, special rules apply: E.g. constants are multiplied by just multiplying their values. If compatible (based on start_time, dt, and number of samples), piecewise constants are multiplied by multiplying their sample arrays. If no special rules apply, the envelopes are multiplied as functions (e.g. any multiplication involving a generic `Signal`.)

### Additional review notes

I've updated all other files that used `Signal` and its subclasses, along with all existing tests, which are now passing. I've also updated the example notebooks so they all run with the current interface. 

The notebook `new_signals.ipynb` demos the new interfaces and functionality. 

Still to do:
- Add new tests for new functionality (especially for vectorized evaluation). 
- To get things running I've had to do a bit of `Array` hackery again. It might be good to start putting explicit jax implementations of some components to avoid this as much as possible. Should we just check the default backend and use jax logic if it's `'jax'`?
- Once things are settled I'll update the original/merge the original `Signals.ipynb` demo with some new stuff in `new_signals.ipynb` to finish off the PR.

I haven't implemented absolutely everything we discussed, but I think at this point, if we're happy with this structure, we can leave subsequent iterative changes for subsequent PRs that can be worked on in parallel with other things that depend on the `Signal` structure.

DanPuzzuoli commented:
To do since last discussion (crossed out means completed):

- Rename PiecewiseConstant -> DiscreteSignal
    - Same with PiecewiseConstantSignalSum
- ~Remove `np.vectorize` option from Signal - put in the docs that this is assumed~
- Plotting (may leave this for the end)
    - Return axis object?
    - Defaults: add a grid
    - Defaults: Add legend for real/imaginary where relevant
    - Rename plot -> draw
    - Have a single draw function and just specify which via a “method” kwarg or something like this
- repr
    - Make it look like the default python repr (with angle brackets and what not) but a bit more concise
- Change collapse -> flatten or reduce
- Change to_pwc to discretize
    - Make it a class method of Discrete signal and of DiscreteSignalSum
- Properly flesh out `SignalList` object and update locations that used `VectorSignal`
- Write implementation/test everything with jax!!!
    - Can we use dispatch default backend to choose jax behaviour?
- Add everything to docs

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -27,167 +28,196 @@
-            n_samples: number of steps to resample with.
-            start_time: start time from which to resample.
+    @carrier_freq.setter
+    def carrier_freq(self, carrier_freq: float):
+        """Carrier frequency setter."""
+        if type(carrier_freq) == list:
+            carrier_freq = [Array(entry).data for entry in carrier_freq]
```
Something seems inconsistent here between the type `carrier_freq: float` and the immediate list check.

DanPuzzuoli replied:
Ah yeah, right. I had added the list check to support subclasses `SignalSum` and `DiscreteSignalSum`, where `carrier_freq` and `phase` are either lists/Arrays. I could add `list` as an acceptable type here but maybe it makes more conceptual sense to just have `SignalSum` directly implement the list/array versions of these getter/setters.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -27,167 +28,196 @@
-    def envelope_value(self, t: float = 0.0) -> complex:
-        """Evaluates the envelope at time t."""
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
```
Is this needed if you set them immediately after? Or does lint or the editor complain?

DanPuzzuoli replied:
Just removed these lines and the linter doesn't complain, so I'll get rid of them!

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -27,167 +28,196 @@
-            dt, samples, start_time=start_time, carrier_freq=self.carrier_freq, phase=self.phase
-        )
+    @phase.setter
+    def phase(self, phase: float):
+        """Phase setter."""
+        if type(phase) == list:
+            phase = [Array(entry).data for entry in phase]
```
Same comment as above.

DanPuzzuoli replied:
Same reply: moved the list handling version to `SignalSum`.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -27,167 +28,196 @@
 
-    def __radd__(self, other):
-        return signal_add(self, other)
+    def __str__(self) -> str:
+        """Return string representation."""
+        if self.name is not None:
+            return str(self.name)
```
Would it perhaps not be nicer to return
```
Signal(carrier_freq={freq}, phase={phase}, name={name})
```
if it has a name?

DanPuzzuoli replied:
My thinking was that if it has a name it should just be displayed to make it more succinct, but I could go either way on this.

eggerdj replied:
Ok. We can leave as is for now. It is easy to change later on and will not have a big impact.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -27,167 +28,196 @@
-            axis (AxesSubplot): the axis to use for plotting.
-        """
-        x_vals = np.linspace(t0, tf, n)
+    def __neg__(self) -> "SignalSum":
+        return -1 * self
```
Does this have to be a `SignalSum`? Could it not be a `Signal`?

DanPuzzuoli replied:
This does actually return a `SignalSum` - the expression `-1 * self` will result in a call to `self.__rmul__`, which returns a `SignalSum`.

eggerdj replied:
Ok, sounds good.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -247,8 +279,53 @@ def __init__(
+            dt: Time increment to use.
+            n_samples: Number of steps to resample with.
+            start_time: Start time from which to resample.
+            sample_carrier: Whether or not to keep the carrier analog or include it in the
+                            sampling.
```
`keep the carrier along` is a bit vague

DanPuzzuoli replied:
Updated this, as well as the corresponding text in `DiscreteSignalSum.from_SignalSum` to be more explicit.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -247,8 +279,53 @@ def __init__(
+            DiscreteSignal: The discretized ``Signal``.
+        """
+
+        times = start_time + (np.arange(n_samples) + 0.5) * dt
```
Perhaps add in the docstring what sampling methodology we use. Here, it appears to be at the mid-point.

DanPuzzuoli replied:
Added this.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -282,24 +359,27 @@ def start_time(self) -> float:
 
     def conjugate(self):
-        return PiecewiseConstant(
+        return DiscreteSignal(
             dt=self._dt,
             samples=np.conjugate(self._samples),
             start_time=self._start_time,
             duration=self.duration,
-            carrier_freq=-self.carrier_freq,
+            carrier_freq=self.carrier_freq,
             phase=-self.phase,
         )
```
How do we interpret `conjugate` for a real signal?

DanPuzzuoli replied:
This is a good point:
- In this PR, for a signal `Re[f(t)exp(i(wt + phi))]`, conjugation maps the signal to one implementing the function `Re[conj(f(t))exp(i(wt - phi))]`; basically taking the conjugate of the envelope, and conceptually including the phase as part of that. 
- I chose this as it is useful in implementing multiplication of signals, as we do really need to just take the conjugate of the envelope/phase. Also, since signals are explicitly real now, also taking `w` -> `-w` would result in the same real-valued function.
- However, this behaviour is definitely weird with the method `complex_value`, as with this definition `sig2 = sig1.conjugate()` won't satisfy `sig2.complex_value(t) == np.conj(sig1.complex_value(t))`.

Some options: 
- We could change `.conjugate()` to `.conjugate_envelope()` to be explicit about this. 
- Or `.conjugate()` could be expanded to have optional arguments about which parameters in the complex representation are being "conjugated". E.g. it could have a signature: 
```
def conjugate(self, envelope=True, carrier_freq=True, phase=True):
```

My preference would be the latter option.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
 
-    .. math::
+class SignalSum(Signal):
+    """Class representing a sum of ``Signal`` objects."""
```
I think this requires more details. Perhaps a few words on its use and how it is implemented.

DanPuzzuoli replied:
Updated.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        """Initialize with a list of Signal objects through ``args``.
 
+        Args:
+            args: ``Signal`` subclass objects.
```
Why not call this `signals`?

DanPuzzuoli replied:
Changed to `signals`

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+            phases.append(sig.phase)
+
+        # set carrier and phase
+        self.carrier_freq = carrier_freqs
+        self.phase = phases
```
So this explains the list check in the setter. I think the type hints need to be changed in `Signal`.

DanPuzzuoli replied:
Changed them in `Signal`.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        self.phase = phases
+
+    def envelope(self, t: Union[float, np.array, Array]) -> Array:
+        """Evaluate envelopes of each component. For vectorized operation,
+        last axis indexes the envelope, and all proceeding axes are the
+        same as the ``t`` arg.
+        """
```
This a bit unclear.

DanPuzzuoli replied:
I've updated this. I've actually removed most of the description - what I was describing was basically what "vectorized" means, which maybe isn't actually necessary here.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        same as the ``t`` arg.
+        """
+        # to do: jax version
+        # not sure what the right way to do this is, here we actually need
+        # to get it to use np/jnp
```
Can you some how do something like checking if jax is installed and then point to it? E.g. a bit like what is often done with `HAS_MATPLOTLIB`?

DanPuzzuoli replied:
yeah - I think the way we need to do this is to just look at `jax.default_backend()`. Something we need to be careful about though is what types we expect the user-defined envelopes to return. (i.e. do we demand they return an `Array` if using jax?)

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        # to get it to use np/jnp
+        return np.moveaxis(Array([env(t) for env in self._envelopes]), 0, -1)
+
+    def complex_value(self, t: Union[float, np.array, Array]) -> Array:
```
docstring?

DanPuzzuoli replied:
Added.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        return default_str
 
-        \Omega(t)*exp(2\pi i \nu t)
+    def flatten(self) -> Signal:
+        """Merge into a single ``Signal``. The output frequency is given by the
+        average.
+        """
```
Would it make sense to have the average frequency as the default but then let the user override this? For example, with a mixer I probably do not want to have the average as the carrier frequency. We could also think of methods to have the carrier at the frequency where the power spectral density of the signal is highest.

DanPuzzuoli replied:
I agree, though I think this requires enough thought that it warrants its own subsequent PR. Currently I'm imagining:
- `flatten` is a brute force function that forces a sum into a single `Signal`, and should be avoided where possible. I view this mainly as a stopgap for the near-term until we update the rest of the repo to work with `SignalSum` objects.
- I think we should add a `simplify` method (in a subsequent PR) that more cleverly attempts to merge signals in a sum based on either their PSD, or on their `carrier_freq` values (and we can have a tolerance here).

I think for AWG implementation we want to use `simplify`, so that it doesn't merge components that are too far apart in frequency space.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
-    Raises:
-        Exception: if signals cannot be added.
+class DiscreteSignalSum(DiscreteSignal, SignalSum):
+    """Represents a sum of piecewise constant signals, all with the same
+    time parameters: dt, number of samples, and start time.
```
```suggestion
    """Represents a sum of discretized signals, all with the same
    time parameters: dt, number of samples, and start time.
```

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        phases: float = None,
+        name: str = None,
+    ):
+        """Samples array has 0th axis corresponding to a signal, 1st axis corresponding to samples
+        for each signal."""
```
This needs more detail.

DanPuzzuoli replied:
Updated this doc string.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
-        phases: Array,
-        drift_array: Optional[Array] = None,
+    @classmethod
+    def from_SignalSum(
```
Does this need to do any validation to check the assumptions of DiscreteSignal sum or does it enforce them by sampling the given `SignalSum`? Perhaps this can be clarified in the docstring?

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
-            Array: the signal envelope at time t
-        """
-        return self.envelope(t)
+    def __init__(self, signal_list: List[Signal]):
+        self.components = signal_list
```
Docstring

DanPuzzuoli replied:
Added.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
-            -self.phases,
-            np.conjugate(self.drift_array),
+    @property
+    def drift(self) -> Array:
+        """Return the 'drift' Array, i.e. the constant part of the ``SignalList``."""
```
By `constant part` do you mean the sub-list of all constant signals? Can you clarify? So if understand this part it will allow us to conveniently separate drift and control parts of a Hamiltonian?

DanPuzzuoli replied:
I've updated this to try to be more clear.

And yes, you're correct, the ultimate use case is to be able to determine the "constant part" of a model (like a Hamiltonian).

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+
+        return Array(drift_array)
+
+    def __getitem__(self, idx: Union[int, List, np.array, slice]) -> Signal:
+        """Enables numpy-style subscripting, as if this class were a 1d array."""
```
This looks like a copy of `__getitem__` from `SignalSum`. It might make sense to introduce an abstract base class to remove some of this redundency.

DanPuzzuoli replied:
Yeah possibly - we could abstract out `__len__` as well. I'll think a bit, but it's probably worth it just for `__getitem__` alone since it's so long.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        return len(self.components)
+
+
+def signal_add(sig1: Signal, sig2: Signal) -> SignalSum:
```
To keep things more modular it might make sense to put these type of functions in a `signals_operations.py` file.

DanPuzzuoli replied:
Just tried but got circular import issues, so I think we'll need to keep these together.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+    try:
+        wrapped_sig1 = to_SignalSum(sig1)
+        wrapped_sig2 = to_SignalSum(sig2)
+    except QiskitError as e:
```
Does our lint allow variables name as a single letter?

DanPuzzuoli replied:
No but I disabled name checking for the whole file this as the argument name `t` doesn't conform to this. Updated though to be called `qe`.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        - If two ``DiscreteSignal``s have compatible parameters, the resulting signals are
+        ``DiscreteSignal``, with the multiplication being implemented by array multiplication of
+        the samples.
+        - Lastly, if no special rules apply, the two ``Signal``s are multiplied generically via
+        multiplication of the envelopes as functions.
```
What happens to carrier phase and frequency?

DanPuzzuoli replied:
Updated comment to explain.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+    return prod1 + prod2
+
+
+def sort_signals(sig1: Signal, sig2: Signal) -> Tuple[Signal, Signal]:
```
What does this enable?

DanPuzzuoli replied:
This is just to enable more succinct type checking for binary operations on signals that care about special cases. E.g. multiplication has a special rule for if one `sig1` and `sig2` is a `Constant` and the other is a more general `Signal`. By sorting `sig1` and `sig2` first, the multiplication routine only needs to check the single case: 
```
type(sig1) is Constant and type(sig2) is Signal
```

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+    """
+
+    if isinstance(sig, (int, float, complex)) or (isinstance(sig, Array) and sig.ndim == 0):
+        sig = Constant(sig)
```
Is `Constant` a `SignalSum`? Does this work with the error raised below?

DanPuzzuoli replied:
The subsequent check wraps any subclass of `Signal` that is not a `SignalSum` in a `SignalSum`, so this will handle the `Constant` case.

eggerdj reviewed (changes requested):
Looking good. See comments. It might also be worth it to
* Add a legend to the plot based on the signal names if defined.
* Add the name of the signal to the `__repr__`

haggaila commented:
I have two small questions:
1. When `conjugate`'ing, shouldn't the conjugate of exp{i v t + phi} negate v as well (in addition to phi)?
2. In the new_signals.ipynb demo, the plots of the discretized signals have a slope between changing values, which is not the way I understand a discrete signal (which must jump sharply). Could it be an artifact of the plotting?

DanPuzzuoli commented:
@haggaila 

> I have two small questions:
> 
> 1. In the documentation above it is not mentioned that `conjugate()`, negates the frequency as well (like the code does).

`.conjugate()`  _used_ to negate frequency as well, but this PR has changed it so that it only conjugates the envelope/phase. I'm going to change it back to also negating the frequency though, based on your and @eggerdj 's comments. (I had a reason for changing this, but I'm realizing now that for the application I had in mind it doesn't make a difference, and your comments have made it clear that it is kind of confusing.)

> 2. In the new_signals.ipynb demo, the plots of the discretized signals have a slope between changing values, which is not the way I understand a discrete signal (which must jump sharply). Could it be an artifact of the plotting?

No this is the correct behaviour. This is arising in plots of `DiscreteSignal`s in which only the envelope has been discretized (so the carrier wave is still analog). The notebook also shows instances of discretizing the carrier wave as well: when plotted they are clearly fully piecewise constant. It's useful to have both behaviours.

DanPuzzuoli commented:
I've made updates based on suggestions. @eggerdj I've replied to your comments - either directly implementing your suggestions or giving some explanation as to what is going on.

Other changes of note:
- Created `SignalCollection` class for abstracting the internal list handling (`__len__`, `__getitem__`, etc) from `SignalSum`, `SignalList`, and `DiscreteSignalSum`.
- Updated `draw` to have a title/legend. The title is based on which function is being + the `str(self)`, and the legend shows Real/Imag when plotting complex data.
- JAX/`Array` handling
   - I've changed all of the signal evaluation functions (`envelope`, `complex_value`, and `__call__`) to only do `Array` wrapping when the backend is `jax`. This way, there shouldn't be any `Array` overhead when using straight numpy. I think maybe this can be used as a "general principle" when coding with the `dispatch`: for functions that contain both numpy/jax implementations, the return type can be `Union[np.array, Array]`, with the presumption that `numpy` implementation may return an `np.array`, and `jax` implementation (or another if added) will return an `Array`. This makes the type hinting a bit less concise but I do think it's perhaps "natural" in that, obviously, the `Array` wrapping is totally unnecessary for the pure `numpy` implementation.
   - In relation to the above, there is now a helper function at the end of `signals.py`, `array_funclist_evaluate`, which sets evaluation of a list of functions for which `Array` wrapping is necessary. `SignalSum` and `SignalList` both use this to setup how they evaluate `envelope`, `complex_value`, and `__call__` when operating in JAX-mode. This function is super small/simple, but it's useful to abstract this out into a helper function as it needs to occur in multiple places.

Question @eggerdj : 
- For `__repr__`, did you mean to actually implement `__repr__` or was that question related to `__str__`? I replied to your comment about `__str__` that I want the `name` attribute to override the behaviour of `__str__` as a mechanism for making the printing of `Signal` names a bit more concise. Thoughts?

haggaila commented:
I have a small suggestion, to have two setters and getters for the carrier frequency in `BaseSignal`, one for frequency and the other for angular frequency (with the 2 \pi factor). To backup the necessity of this - backends report frequencies in both formats (and even for the same physical variable);
For example the qubit anharmonicity is a frequency in `backend.properties().qubit_property(1, 'anharmonicity')`, but an angular frequency in `backend.configuration().hamiltonian['vars']['delta0']`.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +447,693 @@ def add_samples(self, start_sample: int, samples: List):
+        if type(idx) == np.ndarray and idx.ndim > 0:
+            idx = list(idx)
+
+        sublist = None
```
```suggestion
```

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +447,693 @@ def add_samples(self, start_sample: int, samples: List):
-        carrier_freqs = Array(
-            [Array(getattr(sig, "carrier_freq", 0.0)).data for sig in signal_list]
+        freq = signal_sum.carrier_freq
+        samples = None
```
```suggestion
```

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +447,693 @@ def add_samples(self, start_sample: int, samples: List):
         )
 
-        phases = Array([Array(getattr(sig, "phase", 0.0)).data for sig in signal_list])
+    def complex_value(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
```
docstring?

DanPuzzuoli replied:
added

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +447,693 @@ def add_samples(self, start_sample: int, samples: List):
+        dt: float,
+        samples: Union[List, Array],
+        start_time: float = 0.0,
+        duration: int = None,
```
It does not seem like this class is using duration. Is this needed here?

DanPuzzuoli replied:
Yeah it isn't ever used. It only serves as a possible convenience argument for specifying a `DiscreteSignal` of a given length where all samples are `0`. I'll remove it for now and we can see if we ever end up actually needing that.

eggerdj commented on test/ode/signals/test_signals.py:
```diff
@@ -17,213 +17,843 @@
+        self.assertAllClose(discrete_conj.dt, self.discrete2.dt)
+
+
+class TestSignalSum(QiskitOdeTestCase):
```
Perhaps add a test to check that `self.sig_sum1 - self.sig_sum2` behaves as expected?

DanPuzzuoli replied:
Added `self.double_sig_sum = self.sig_sum1 - self.sig_sum2` in the `setUp` and then added the equivalent checks for `self.double_sig_sum` in each test case.

eggerdj replied:
Thanks!

eggerdj reviewed (changes requested):
Looking good. I noticed a few minor things.

DanPuzzuoli commented on qiskit_ode/signals/signals.py:
```diff
@@ -11,183 +11,230 @@
-    envelope function and a carrier frequency.
-    """
+        def conj_env(t):
+            return np.conjugate(self.envelope(t))
```
Yeah: in the `__init__` I've set it so that if jax is being used, the output of the envelope function will always be wrapped in an `Array`, so here the dispatch will implement `np.conjugate` via `jnp.conjugate`.

eggerdj replied:
Thanks.

DanPuzzuoli commented on qiskit_ode/signals/signals.py:
```diff
@@ -11,183 +11,230 @@
-    @abstractmethod
-    def envelope_value(self, t: float = 0.0) -> complex:
-        """Evaluates the envelope at time t."""
+    Note: this class assumes that the envelope function is vectorized. If it isn't, it can
+    be vectorized automatically by calling ``numpy.vectorize``\, or, if using JAX, by
+    calling ``jax.numpy.vectorize``\.
```
Added.

DanPuzzuoli commented on qiskit_ode/signals/signals.py:
```diff
@@ -196,23 +243,28 @@ def __init__(self, value: complex, name: str = None):
+        self._name = name
+        self._value = np.real(Array(value))
+        self.phase = 0.0
+        self.carrier_freq = 0.0
```
Hmm interesting. Yeah it works, changed it to this. I also removed some of the methods as they become redundant after making this change. This has left `Constant` pretty much redundant. 

Should we just get rid of `Constant`? I liked having it be obviously different from `Signal`, but given that the class is basically empty, maybe that's not a good enough reason. I think one potential way of doing this is to add an `is_constant` property to `Signal` (set to `True` if the envelope is specified as a number with 0 freq), which can be used to check if something is constant (as opposed to the type checks that check for `Constant` instances).

Thoughts?

haggaila replied:
I can't see any argument in favor of one or the other..

DanPuzzuoli replied:
I think I may try getting rid of `Constant`. My only initial resistance to it was that I didn't like the idea of a user specifying `Signal(1.)` to represent the constant `1.` - the usage of the word "signal" when setting a constant seemed weird. I think though that we can just completely eliminate the need for the user to ever construct a `Signal` object for constants: they can just leave them as numbers (floats, complex, whatever), and the conversion to `Signal` objects that are constant can all happen behind the scenes. This already occurs for addition/multiplication: if you do `4 + signal`, the `4` is automatically converted to a `Constant`. The only other place this auto conversion needs to be added is in `SignalList` (relevant for specifying model coefficients).

DanPuzzuoli commented:
> I have a small suggestion, to allow setting and getting the carrier frequency both as a frequency and as an angular frequency (so the user wouldn't have to mess with the 2 \pi factor). To backup the necessity of this - backends report frequencies in both formats (and even for the same physical variable);
> For example the qubit anharmonicity is a frequency in `backend.properties().qubit_property(1, 'anharmonicity')`, but an angular frequency in `backend.configuration().hamiltonian['vars']['delta0']`.
> However, I see that currently the frequency can only be set in Signal, and only via the constructor, so it could be complicated to support both.

Yeah I think for now I'd rather keep it simpler and just have one way of specifying it. If it ends up being annoying we can change it later.

DanPuzzuoli commented:
@eggerdj 

ready for re-review

eggerdj commented on test/ode/signals/test_signals.py:
```diff
@@ -17,213 +17,943 @@
+        )
+
+
+class TestDiscreteSignalSum(TestSignalSum):
+    """Tests for DiscreteSignalSum."""
+
+    def setUp(self):
+        self.signal1 = Signal(np.vectorize(lambda t: 0.25), carrier_freq=0.3)
+        self.signal2 = Signal(lambda t: 2.0 * (t ** 2), carrier_freq=0.1)
+        self.signal3 = Signal(lambda t: 2.0 * (t ** 2) + 1j * t, carrier_freq=0.1, phase=-0.1)
+
+        self.sig_sum1 = DiscreteSignalSum.from_SignalSum(
+            self.signal1 + self.signal2, dt=0.5, start_time=0, n_samples=10
+        )
+        self.sig_sum2 = DiscreteSignalSum.from_SignalSum(
+            self.signal2 - self.signal3, dt=0.5, start_time=0, n_samples=10
+        )
+
+        self.signal1 = DiscreteSignal.from_Signal(self.signal1, dt=0.5, start_time=0, n_samples=10)
+        self.signal2 = DiscreteSignal.from_Signal(self.signal2, dt=0.5, start_time=0, n_samples=10)
+        self.signal3 = DiscreteSignal.from_Signal(self.signal3, dt=0.5, start_time=0, n_samples=10)
+
+        self.double_sig_sum = self.sig_sum1 - self.sig_sum2
```
Is this used anywhere?

DanPuzzuoli replied:
Yeah - it inherits from `TestSignalSum`, so all of the test cases from that class are being run on these signals.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -11,216 +11,262 @@
+        t0: float,
+        tf: float,
+        n: int,
+        function: Optional[str] = "signal",
```
This is a bit of a nit-pick but I'm not convinced by the name `function` here. I don't have a good alternative to offer either. Perhaps, `element` or `component`?

DanPuzzuoli replied:
Yeah I'm not crazy about the name `function` either, but I've also been unable to come up with a better alternative. I don't like `element` or `component` as I think there is a bit of naming collision with `SignalSum`/`DiscreteSignalSum`, which actually have elements/components. 

The only other alternative I can think of that seems okay is `method`. 

- Pro: this arg is literally choosing which method (in the class sense) that is being plotted 
- Con: "method" sounds like it's referring to a "plotting method"

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -11,216 +11,262 @@
 
-    def conjugate(self):
-        return Constant(self._value.conjugate())
+        plotter = plt
+        if axis:
+            plotter = axis
```
```suggestion
        plotter = axis if axis else plt
```

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -231,24 +277,75 @@ def __init__(
-            self._samples = Array(samples)
-        else:
-            self._samples = Array([0.0] * duration)
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
+        self._carrier_arg = None
+        self._phase_arg = None
```
Could this not be done in the base class? It feels strange to not call `super().__init__` here but I see why you did it since `Signal.__init__` wants to define a `self._envelope` which `DiscreteSignal` does not have. Can we not bring these classes closer together? E.g. by renaming `self._samples` to `self._envelope` in `DiscreteSignal` and letting `Signal.__init__` accept an `Array` as well as `Callable, complex, float, int` (you already do a conversion in the `__init__` of signal)

DanPuzzuoli replied:
I agree in general with this comment, though I've tried a different approach than your suggestion: 

- I've instead changed `DiscreteSignal.__init__` to construct the envelope evaluation function (basically just the code that was in `DiscreteSignal.envelope`), then call `Signal.__init__` with this envelope. 
- Following this, I changed `DiscreteSignalSum.__init__` to call `DiscreteSignal.__init__`, as the envelope evaluation functions are the same in both of these cases.
- I also changed `SignalSum.__init__` to construct the envelope function (again, from code that was originally in `SignalSum.envelope`), then again called `Signal.__init__` with this envelope (as in `DiscreteSignal`).
- As a result of the above changes I was able to eliminate:
   - Several redundant lines of code in various `__init__` functions   
   - Overriding `envelope` methods in `DiscreteSignal` and `SignalSum`.
   - Unrelated to inheritance changes, but I realized the `complex_value` method in `DiscreteSignalSum` was redundant, as it inherits the correct form of this method from `SignalSum`, so this was also removed.

I think this might clean things up now - everything calls back to the original `Signal.__init__`, with array evaluation logic set up in the respective inits.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -231,24 +277,75 @@ def __init__(
+            samples = signal.envelope(times)
+
+        return DiscreteSignal(
+            dt, samples, start_time=start_time, carrier_freq=freq, phase=signal.phase
```
Why no name?

DanPuzzuoli replied:
Not sure what name to assign to this:

- If the original signal has no name then I think it's fine to not assign a name.
- If the original signal does have a name, if we want to assign a name we could do something like `'Discretized(dt=0.1, name)'`? Or do you think we should just set the name to the name of `signal`?

eggerdj replied:
I think we can just set the name to the name of the signal.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +433,696 @@ def add_samples(self, start_sample: int, samples: List):
+            elif isinstance(sig, (int, float, complex)) or (
+                isinstance(sig, Array) and sig.ndim == 0
+            ):
+                components.append(Signal(sig))
```
Seems like `Signal` should accept `Array` as input too? See my comment above.

DanPuzzuoli replied:
See response to above comment.

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -11,216 +11,262 @@
-
-        Returns:
-            A piecewiseConstant signal.
+            envelope: Envelope function of the signal, must be vectorized.
```
From the docstring I would expect init to raise an error if the envelop is not vectorized but I cannot find one. Why?

DanPuzzuoli replied:
I don't believe it's possible to check for this at instantiation. I think the only way to really find out that something _isn't_ vectorized is if it fails to produce arrays of the correct shape on some input. I'm not sure what we could do here aside from adding checks in every signal evaluation function that the input shape = output shape, but this seems like unnecessary overhead. I'm tempted to not worry about this for now: if something isn't vectorized I believe it should result in pretty clear errors (of course, I could be wrong).

eggerdj commented on qiskit_ode/signals/signals.py:
```diff
@@ -11,216 +11,262 @@
+
+    def __init__(
+        self,
+        envelope: Union[Callable, complex, float, int],
```
Can this not also take an `Array` as input? See my comment below on `DiscreteSignal`. The if-else structure in this init also suggests the we should accept Array.

DanPuzzuoli replied:
See response to above comment.

eggerdj reviewed (changes requested):
Almost there. I think we might be able to better align the internals of `Signal` and `DiscreteSignal`

eggerdj reviewed (changes requested):
A few more small comments.

DanPuzzuoli commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +427,678 @@ def add_samples(self, start_sample: int, samples: List):
+        start_time: float = 0.0,
+        carrier_freq: Union[List, np.array, Array] = None,
+        phase: Union[List, np.array, Array] = None,
+        name: str = None,
```
good catch, added it to the call to `DiscreteSignal.__init__`

DanPuzzuoli commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +427,678 @@ def add_samples(self, start_sample: int, samples: List):
-    Raises:
-        Exception: if signals cannot be added.
-    """
+        Signal.__init__(self, envelope, carrier_freqs, phases)
```
I assume this is solved with the type hints in your other suggestions

DanPuzzuoli commented on qiskit_ode/signals/signals.py:
```diff
@@ -327,202 +427,678 @@ def add_samples(self, start_sample: int, samples: List):
+        if phase is None:
+            phase = np.zeros(samples.shape[-1], dtype=float)
+
+        DiscreteSignal.__init__(self, dt, samples, start_time, carrier_freq, phase)
```
I assume this is also solved by the type hints in your comments

DanPuzzuoli replied:
Not sure if I messed up committing one of your suggestions, but I just manually added the updated type hints/doc description for `DiscreteSignal` - which should resolve this.
